Certbot for Windows auto-update mechanism

adferrand opened the pull request:
This PR implements a mechanism to make Certbot for Windows update itself during the scheduled renew task run twice a day. 

Technically this PR blocks on the signature of the installer executables, but all of this will be added soon, so we can start the review. 

The mechanism is written into pure Powershell. I did that for the following reasons:
* other programs in the wild usually use an approach of a dedicated executable outside the program itself
* because Windows is Windows, it is just impossible for a program to self update: any opened file descriptor could block the process. So at least a part of the logic needs to be outside Certbot. Some can be in it (like checking new version availability, downloading the installer), but it was a minor part, and having the logic splitted in two different languages would decrease the maintainability in my opinion
* Powershell includes all high level features required to make the script short and readable
* `exec` Linux style is not possible on Windows, so a process (like Certbot) cannot transform itself into another process (like the upgrade mechanism), so a wrapper on top of Certbot would be required. With the limitations of Windows I do not think it would make something very elegant. 
* this way the mechanism can be plugged freely anywhere. I think it is relevant to associate it to the scheduled renew task, to avoid to break it if the upgrade would happen in parallel to `certbot renew` call. Here it is not possible, because the renew and the upgrade are called sequentially one after the other. 

As said, the upgrade script is plugged into a specific action in the existing scheduled task. Actions in a task are run sequentially by construction. Also, it is not the script itself that is run, but a copy of it outside Certbot install directory, here again to avoid errors when trying to delete an open file during the upgrade. 

The script itself will:
* get current version of Certbot installed locally
* get remote latest version using the Github API
* download if necessary the installer of this new version
* check it is signed correctly
* check it is signed by us
* run the installer in silent mode, using the original installation directory specified by the user when Certbot has been installed the first time

Potential improvements : check that Certbot is effectively installed with the new version, fallback to previous version otherwise.

bmw commented:
I'm not going to have the time to give this a full review before my few days off, but I had some high level questions/suggestions which could get us started on the review here.

The first is that I'm tempted to actively avoid Powershell. No one other than you on the Certbot team currently knows the language and being a bunch of UNIX nerds, I doubt many of us would be excited to learn it for better or worse. If you disappeared tomorrow, I think we'd struggle to maintain any Powershell code (at least initially). Due to problems we've talked about before, I don't think we can avoid it entirely, but I'm tempted to try to push as much of it into Python as we reasonably can to reduce the amount of (new) technology people have to understand to maintain Certbot.

If you think that's a reasonable idea, I have a new idea on how we could maybe do this that I've tested. The idea is to spin up a new process to handle running the installer. I tested this by writing the simple script:
```
import subprocess
subprocess.Popen([r'C:\Users\IEUser\certbot-beta-installer-win32.exe', '/S'])
```
and then running:
```
'C:\Program Files (x86)\Certbot\Python\python.exe' script.py
```
The process started by `Popen` outlives the Python script that started it.

Potential downsides with this approach I've thought of are:

1. You have to manage concurrency between the Python code and the process it starts.
2. Error handling/reporting may be worse.

My other questions were around when Certbot updates and its stability:

1. Did you decide that Certbot trying to update itself at the beginning or end of every execution was too slow?
2. This can maybe come later, but what if someone tries to run Certbot while the updater is running?
3. I'm not sure we can be totally safe from renewal failing because an update is running by running them sequentially with our current setup. Because Certbot is scheduled to run at a random time every 12 hours, isn't it possible that the scheduled task runs at the end of the first 12 hour window and the beginning of the next and they overlap when taking into account the time it takes to run Certbot and update it?

adferrand commented:
So last days I studied and thought a lot about the best approach to implement an automated upgrade process of Certbot under Windows.

TL;DR: I am strongly convinced that deporting the upgrade logic in an independant process from Certbot is the best move. I am open to modify the actual technologies involved if it can reduce the technological risk.

As said initially, having a separated process to upgrade is nothing new on Windows. A lot of pieces of software use this approach, with major editors involved, like Adobe, Microsoft, Google. Some of them use dedicated services, other dedicated scheduled tasks in the Task Scheduler.

I obviously cannot foresee all reasons that led the engineers behind these implementations to choose that approach. But I am tempted to trust them, since they are far more experienced in Windows development than me, and they followed the chosen approach for years. However, I may anticipate some reasons:
* first one is readability of the process: using services or scheduled tasks, the process output (logs and results) is consumed by Windows and made available in a consistent way. I would expect an experienced Windows administrator to know how to debug things if something goes wrong with that approach
* second is that having the upgrade separated from the program itself makes the process more robust, in particular in case of critical failure in the program itself.

Let's elaborate.

It has to be noted that several programs ship their upgrade right into the main program. I see that all these programs have a common pattern: they are UI tools (like visual editors, productivity suite) explicitly launched by the user.

In that context here is how an upgrade would happen: the user opens the program, and receives a pop-up informing him about a new upgrade. The user is given the typical choice to execute right now the upgrade, or postpone it until the end of current program execution. Once the upgrade process is started, a new UI is opened (the installer one) that displays progress information, and allows the user to acknowledge the success or failure of the upgrade.

Certbot, on the other hand, is a CLI tool, whose execution is mainly done non interactively. Basically goal of Certbot is to configure the required certificates, then forget about it, since the renewal process will be handled in the background. Certbot failures are seen asynchronously, like 3 months after when certificates start to expire because the renewal process failed for any reason. This situation is quite different.

First, since Certbot actions are mainly non-interactive, having a trace of upgrade process available in the Windows logging system is desirable. But one can argue that this could be logged in the main log file of Certbot (typically C:\Certbot\logs\letsencrypt.log).

However, here is about my second point: your proposition @bmw would be about having (a part of) the upgrade process inside Certbot. Indeed playing around the subprocess capabilities can be done, but it is not the main problem for me. Because here what is bother me is the idea of associating the upgrade process with Certbot itself: if for any reason (and there is a lot in Windows) Certbot is broken, then the upgrade process is also, we have no way to push automatically an upgrade to fix that. With a dedicated process, we have a one tool one purpose logic, and a tool more likely capable of fixing errors automatically (user errors, Windows-magic runtime errors or developers errors).

I understand the interest to push it in Certbot, because it gives a natural way to implement most of the logic in Python, but this technological risk mitigation alone is not worth it compared to the risks and benefits describe above about having a dedicated upgrade mechanism.

That said, we can focus now on how to implement that, and see the technological risks, since new technologies to be added. I see three of them: using the Task Scheduler, using Powershell, using the  .NET framework.

On the first one, I think we do not add a risk in fact: Task Scheduler is already used by the certbot renew task. We will obviously avoid implementing the upgrade as a Windows service, since the Task Scheduler has everything needed. We should however avoid to play to much with the Task Scheduler: I particular like adding/removing dynamically one-go new tasks, like in one of my past proposition. Instead we should reuse the existing task and extend it with minimal modifications. This way we can benefit at low cost from the Task Scheduler functionalities: mainly that it adds convenient way to handle concurrency: none other certbot renew + upgrade task can kicks in while another one is currently running.

On the second one, I understand that Powershell is a new language to maintain. I will defend that it is a Windows native approach, and its cost added to the technological complexity is quite low too what we have already. What I mean here is that Certbot for Windows requires already advanced knowledge on how Windows works (I think in particular here about file permissions, and communication to Win32API through pywin32 module). An additional Powershell script is not that much compared to it, in particular when we consider that there are already Powershell scripts in the Windows installer.

On the third one, I am thinking of the calls to .NET in order to check the Authenticode in the current PR. I think this technological risk is unavoidable, since there is no equivalent in Python. These kind of calls will need to be implemented, either as subprocesses of Powershell scripts or Python mapping using pythonnet. It is intended to grows a lot in the future with the IIS plugin. This is another story, but a sensible choice will need to be done on that matter.

So to sum up, what I describe here is quite what is proposed in the PR already. In term of strategy, I think it is the best. In term of technological risks, the use of Task Scheduler and .NET library can not be avoided.

Even if the risk is not high in my opinion, we can discuss about the Powershell script. The best second possibility in my opinion, while respecting the dedicated process principle, would be to use pywin32exe or pywininstaller to have a self-executable Python script.

Last thing is a question of concurrency. If the Task Scheduler prevents to run two certbot renew (and two upgrades) in parallel, there is still the problem of a user that would run Certbot during the upgrade. It is more a corner case, that we can handle later.

bmw commented:
In addition to Google Omaha, https://github.com/vslavik/winsparkle could maybe be an option for us to avoid writing and maintaining our own solution. I'm curious to hear what you think.

bmw commented:
It looks like there are a few options out there. Some others are mentioned in the discussion of https://news.ycombinator.com/item?id=7584520.

I personally think its worth looking into this issue a bit and seeing if there's a good auto-updater framework for Windows we can use.

adferrand commented:
The thread is 5 years old, so I started from it and continued digging from Omaha and WinSparkle that you provided. I will add, as major upgrade framework, also TUF.

I would definitly go for WinSparkle, that is simple, efficient, very well documented, easy to use in Python since it is a simple DLL.

But, I would do it for a software with an UI. It gives you, when you open your program, various UIs to control the upgrade process. Same for Omaha (that we see with Google Chrome) and TUF (that is used by Docker for Windows).

All these products are sadly not a good fit for Certbot, since Certbot is a CLI tool, whose most of its execution will be done non-interactively during the two daily `certbot renew` calls. In this context, these UIs would show up, depending on our choice:
* (_if the upgrade process is tied to interactive calls only_) only when Certbot is called interactively by the user, so it would happen rarely (potentially with several months gap on a production system). And having a UI showing up when you are on the console, is arguably not a super cool user experience
* (_if the upgrade process is tied to any calls_) at any time, any where. The user is working on the machine, then the UI pops up from nowhere ... Or the UI is just not showing up at all, since the user is not logged in.

I think we are basically on our own here, with just the good practices taken from other programs that are in an equivalent situation, as I described in my previous post.

bmw commented:
Just to make sure, you feel like you have taken the time to research existing Windows packaging frameworks and there is nothing we can use? There was only an hour between my comment and yours and we had a meeting in between.

If that's the case, that's fine, but I want to make sure that's what's really going on here and we're not just attached to our existing implementation. I am going to be verifying this as part of the review and you doing the initial research would help me review your work more quickly so I don't have to start from scratch. Some of what you said surprised me because Omaha works by registering a scheduled task to check for updates automatically. I am not aware of it ever prompting you to install an update.

adferrand commented:
It is not about this specific hour, but all the work and research done before and after the creation of this PR 23 days ago. I think we start to have a good perspective of what is offered in this area.

Let's deal with Omaha (and TUF, that is similar) again. The scheduled task is only a small part of the framework, it is designed to offer a full UI experience on a self-upgrading program, starting with a specific installer/upgrader, and after following the life of the product. It involves a client-server architecture, with specific protocol and binaries to operate both side.

Is it really something we want to be entitled with, to ask to EFF ops to maintain that architecture?

Also have a look to the libraries requirements to build Omaha on client side. This is not a reasonable requirement to ask for any developer (particularly Linux-centric ones) on Certbot.

WinSparkle is simpler, just have a look to the header file comments in the project, that explain how it works.

Basically I am not attached on the implementation I proposed here, but on the design I extracted from the specific requirements of Certbot versus other programs. On that perspective the available frameworks does not fit in that design, while the implementation in this PR does.

We would largely prefer starting with this simple implementation, and having something that works now instead of designing for months, and pivotate in the future on a more complex solution, or delegate to other programs, only if this is necessary.

adferrand commented on windows-installer/tasks-up.ps1:
```diff
@@ -0,0 +1,29 @@
+    $adminGroupID = $adminSID.Translate([System.Security.Principal.NTAccount]).Value
+    $principal = New-ScheduledTaskPrincipal -GroupId $adminGroupID -RunLevel Highest
+    Register-ScheduledTask -Action $actionRenew, $actionUpgrade -Trigger $triggerAM, $triggerPM -TaskName $taskName -Description "Execute twice a day the 'certbot renew' command, to renew managed certificates if needed." -Principal $principal
+}
```
Not relevant anymore, but to respond to this, yes, the script is working from command line without it, and adding it did not fixed the problem you mentionned on point 1.

adferrand commented:
> 
> 
> Here's the start of a review here.
> 
> I spent some time looking for a well maintained update framework that'd work well for us and found nothing. Your notes made me verifying that faster so thanks for that. It constantly amazes me how much time we spend working on packaging. If someone was so inclined, I think there's a lot of potential valuable open source projects people could write to help with client side Python/Windows packaging.
> 
> I do not totally agree with everything you wrote in [#7539 (comment)](https://github.com/certbot/certbot/pull/7539#issuecomment-555534449), but I think rather than getting into the weeds in that, I think it's more productive to focus on what you have here.
> 
> I haven't fully reviewed all files, but in addition to my inline comments, the questions I have are:
> 
>     1. I tried testing this but the installer/uninstaller failed to add or remove the scheduled task. Does it work for you? If not, can you fix it?
> 
>     2. I believe the update process will try to delete the scheduled task while its running. Is this a problem? Are the nice logs/results you described still available after doing this? If so, where?
> 
>     3. I believe the update process will try to delete the update script while it is running. Isn't this a problem or am I misunderstasnding something?
> 
>     4. One of the things you've talked about wanting is a fallback mechanism and that doing this isn't as feasible with other approaches. Can you describe at the highest level how this would work with the approach you have here?

1. Yes it seemed that calling the script from NSIS was not working when the script is defined as a full advanced function (`CmdletBinding`). Since it is not required here, I remove it, and the script is working well again. Note that I keep it for the update script itself, since it seems that Windows call it through a pipeline as a scheduled task, and it removes some weird (but non-critical) logs error at the end of the execution.

2. No, suppressing the task from the called script is not a problem. Logs are stored at `C:\Certbot\log\auto-update.log`, and are kept when Certbot is uninstalled.

3. No this is not a problem, because it is a script, not an executable. On an executable you would have a problem, since the file would still be opened by the kernel to the execution stack. But a Powershell script, as any script, has its content loaded inside the interpreter before instructions are actually run. Once the interpreter is running the command, the script file itself is already closed and so can safely be deleted.

4. My idea for the fallback is:
* to also download the installer for the current version of Certbot installed
* uninstall the old version
* install the new version
* make an assertion about Certbot being fully functional on this new version
Then if not:
* uninstall the new version
* install the old version

adferrand commented on windows-installer/tasks-up.ps1:
```diff
@@ -0,0 +1,29 @@
+    $adminSID = New-Object System.Security.Principal.SecurityIdentifier("S-1-5-32-544")
+    $adminGroupID = $adminSID.Translate([System.Security.Principal.NTAccount]).Value
+    $principal = New-ScheduledTaskPrincipal -GroupId $adminGroupID -RunLevel Highest
+    Register-ScheduledTask -Action $actionRenew, $actionUpgrade -Trigger $triggerAM, $triggerPM -TaskName $taskName -Description "Execute twice a day the 'certbot renew' command, to renew managed certificates if needed." -Principal $principal
```
Done about the description.

About order of actions, I want to renew before upgrading in order to give the additional up to 3 months to the user to fix a broken certbot installation if the upgrade breaks it for any reason.

adferrand commented on windows-installer/tasks-down.ps1:
```diff
@@ -0,0 +1,15 @@
+            Unregister-ScheduledTask -TaskName $task -Confirm:$false
+        }
+    }
+}
```
Not relevant anymore.

bmw commented:
> No, suppressing the task from the called script is not a problem. Logs are stored at C:\Certbot\log\auto-update.log, and are kept when Certbot is uninstalled.

For this question I was referring to what you said in https://github.com/certbot/certbot/pull/7539#issuecomment-555534449 as a benefit of this approach being:

> using services or scheduled tasks, the process output (logs and results) is consumed by Windows and made available in a consistent way.

Are these logs lost when we delete the scheduled task? If not, where can I find them?

I haven't had a chance to review the update script itself yet, but I don't think we should just be putting the log in `C:\Certbot\log` as that path is configurable in Certbot. If Windows has an existing mechanism to store logs/output for us, I think it'd be good to use that.

> No this is not a problem, because it is a script, not an executable. On an executable you would have a problem, since the file would still be opened by the kernel to the execution stack. But a Powershell script, as any script, has its content loaded inside the interpreter before instructions are actually run. Once the interpreter is running the command, the script file itself is already closed and so can safely be deleted.

I don't think this is true of all scripts as certbot-auto had problems with this in the past. See https://github.com/certbot/certbot/issues/2456. Are you confident this isn't a problem in powershell? I think it'd be good to make sure of that to avoid us finding significant problems with the overall approach down the road. Being able to see sources for the behavior you describe would be useful to me.

> My idea for the fallback is:
> * to also download the installer for the current version of Certbot installed
> * uninstall the old version
> * install the new version
> * make an assertion about Certbot being fully functional on this new version
> Then if not:
> * uninstall the new version
> * install the old version

Do you think we could instead keep the existing installation around until we can verify the new one is working?

adferrand commented:
> 
> 
> > No, suppressing the task from the called script is not a problem. Logs are stored at C:\Certbot\log\auto-update.log, and are kept when Certbot is uninstalled.
> 
> For this question I was referring to what you said in [#7539 (comment)](https://github.com/certbot/certbot/pull/7539#issuecomment-555534449) as a benefit of this approach being:
> 
> > using services or scheduled tasks, the process output (logs and results) is consumed by Windows and made available in a consistent way.
> 
> Are these logs lost when we delete the scheduled task? If not, where can I find them?

As basic feature, a scheduled task write high level events through the Windows Event Log API, that is accessible using the Windows Event Viewer. These events are not deleted when the task is deleted, they are part of Windows system itself. However by default you have only which program or command is executed by the task, and what was the exit code.

To store more useful data, the script must write dedicated logs. See below.

> 
> I haven't had a chance to review the update script itself yet, but I don't think we should just be putting the log in `C:\Certbot\log` as that path is configurable in Certbot. If Windows has an existing mechanism to store logs/output for us, I think it'd be good to use that.

Yes, as said above, I am going to switch to this mechanism.

> 
> > No this is not a problem, because it is a script, not an executable. On an executable you would have a problem, since the file would still be opened by the kernel to the execution stack. But a Powershell script, as any script, has its content loaded inside the interpreter before instructions are actually run. Once the interpreter is running the command, the script file itself is already closed and so can safely be deleted.
> 
> I don't think this is true of all scripts as certbot-auto had problems with this in the past. See #2456. Are you confident this isn't a problem in powershell? I think it'd be good to make sure of that to avoid us finding significant problems with the overall approach down the road. Being able to see sources for the behavior you describe would be useful to me.

I could not find any evidence that Powershell is reading script by (large) chunks of data. And the only idea of size outbound limit was about the maximum script size that I found in one unique reference: https://stackoverflow.com/questions/40086313/max-size-of-scriptblock-initializationscript-for-start-job-in-powershell

> 
> > My idea for the fallback is:
> > 
> > * to also download the installer for the current version of Certbot installed
> > * uninstall the old version
> > * install the new version
> > * make an assertion about Certbot being fully functional on this new version
> >   Then if not:
> > * uninstall the new version
> > * install the old version
> 
> Do you think we could instead keep the existing installation around until we can verify the new one is working?

I would prefer to not rely on the local filesystem. If it has been damaged for any reason, the fallback may not work anymore because the installer would not be there. If the system is capable to download the new installer, then it is capable to download the old one. This makes the fallback behavior consistent with the new installation procedure (both in functional and not functional state).

adferrand commented:
The script now use the Event Log engine. One can retrieve logs using the command `Get-WinEvent` -LogName "Certbot auto-update"` on Powershell, or through the Windows Event Viewer tool, in Windows logs > Application (while the later take a couple of minutes before showing the latest events written to the system).

adferrand commented:
About the testing in Powershell 4.0.

If a given testing node has already Powershell >=5.0, it is impossible to make the scripts be run with 4.0 on this node. Indeed, there is a `--version` flag on Powershell, but it works only for trigerring a compatibility mode on PS2.0 or PS3.0. In theory, it is possible to downgrade to Powershell 4.0 a Windows machine that ships natively with 4.0 and have been updated later, by uninstalling the associated KB package. But restarting the machine is requisite to apply the change, as a lot of things on Windows.

Considering that the upgrade script is not going to change a lot once published, by first proposition is that the contributor/reviewer run locally the related scripts on a machine with PR4.0 during a PR.

Considering that installing the most recent Powershell version is a fairly reasonable action, and is in fact proposed in the Windows auto upgrade process, my second proposition is to check Powershell version in the NSIS installer script, and fail the installation procedure if it does not match the requirement (here PS 5.0 since it is the version we can test in the CI pipeline).

adferrand commented:
About testing the Certbot auto upgrade process.

Yes, I think it is relevant to do it know.

Couple of test proposition:
* run directly the script, make various assertions on logs and by calling certbot about whether it installs everything, reinstall if needed (for instance by breaking the current installation), do nothing if nothing is to be done
* install the script as a scheduling task, trigger the task manually, and make the previous assertions through the logging system and call to certbot
* run the installer executable, and proceed to assert the previous bullet point

Last point covers the entire process more as a black box, and has my preference for integration tests types. This could take the form of a self-container Powershell script that can be run on the CI, or locally.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,103 @@
+    #       if ($signature.Status -ne 'Valid') {
+    #           throw "Downloaded installer has no or invalid Authenticode signature."
+    #       }
+    #       if ($signature.SignerCertificate.Thumbprint -ne $installerAuthenticodeCertificateThumbprint) {
```
Do you have a stronger identity check in mind?

adferrand replied:
Like the public key? We will need to always use the same private key, and never loose it over the years though.

bmw replied:
Yeah I think checking the (hash of the) public key probably makes the most sense. We could do something more complicated like what [the update framework](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md) does and have something like an extra key that we use to sign a file specifying the certs/keys that should be trusted, but I personally think that's probably overkill. certbot-auto only trusts a single key.

I think this design is probably something we should flag for everybody in Mattermost or a weekly meeting in case anyone is strongly in favor of something different, but I doubt they will be.

adferrand replied:
I implemented a check against the Certbot public key used for certbot-auto.

bmw commented:
> Considering that installing the most recent Powershell version is a fairly reasonable action, and is in fact proposed in the Windows auto upgrade process, my second proposition is to check Powershell version in the NSIS installer script, and fail the installation procedure if it does not match the requirement (here PS 5.0 since it is the version we can test in the CI pipeline).

If in addition to this we have a check in at least one of the Azure Pipelines nodes that PS 5.0 is being used so we know if we're no longer testing for compatibility, this SGTM. I personally like this better than testing manually.

Can this be done in separate PR that lands before this one?

Also, can we add `#Requires -Version 5.0` to the top of each Powershell script?

> About testing the Certbot auto upgrade process.
> ...

I'm not sure whether you were proposing we implement all of those tests or just a subset of them, but either way is fine with me. I do think having at least some testing of the registered scheduled task is a good idea, but if other testing is cleaner to implement by calling the script directly, I think that's fine too. At a minimum I think we should probably have tests for the two success cases (either nothing needs to be done or we upgrade).

Initially I was going to suggest that any test scripts here are written in Python for familiarity and because I think it's a more robust programming language, but if you're going to be doing much checking event logs, running scheduled tasks, etc., I think you'll essentially have to embed a lot of Powershell scripts in the Python code at which point you might as well just use Powershell.

Another idea I had though is maybe this is a good opportunity to play around with `pythonnet`. It would allow us to write these tests in Python and try the library without shipping it to our users yet. I really don't have a strong opinion on the language or libraries used here, but wanted to share my thoughts.

What do you think?

adferrand commented:
About my propositions of test, the idea was to pick one in priority, and maybe two. Basically I think the last one is the most important, because it tests with the complete logic, but it will also more difficult to troubleshoot if the test fails, while running directly the upgrade script will be easier to troubleshoot but is less significant.

I propose to implement the last bullet point (real installation and manipulation through trigerring of the schedule task), and implement later lower level tests if needed.

About how to write it, yes, I think it is a really good idea to leverage pythonnet. My idea is to write a new integration tests suite in certbot-ci, and add pythonnet as a dependency for this test project.

bmw commented:
Sounds good to me!

bmw commented:
@adferrand, let me know when this is ready for another review.

adferrand commented:
No problem, I keep you informed.

adferrand commented:
@bmw, this PR is ready for a new review. I updated the windows installer integration tests to check that Certbot is able to self-upgrade using its scheduled task.

I chose here a blackbox testing approach: the environment around Certbot is modified, Certbot itself is modified at the bare minimum, and tests are done without knowledge of how Certbot internal logic and as close as possible to the real case.

A test signing certificate is created, and registered properly in Windows to be an acceptable one: a file signed with gets a totally valid Authenticode signature for the duration of the test. A mock of GitHub is setup to trick the upgrade mechanism, and use the dev version of the installer instead of a released one on GitHub. The upgrade mechanism itself consumes two registry keys: one to define what is the acceptable public key for the Authenticode signature to check, one to define which URL to check to get the latest release. Both are used during the test, but can also be used by a Windows administrator for specific needs (like corporate networks).

To check that the upgrade mechanism is working, I take the worst scenario: a broken Certbot installation, and I check that the upgrade repair it.

During the test I also improve the installer and the Powershell script to make them more resilient. For instance, I dropped the multiuser approach in the installer, since we are not using it and it messes up with the ability to define the Certbot installation path in non-interactive scenarios. The upgrade mechanism send more logs, and never break with writting something about it in the logs, and is not dependant to unexpected modifications accessible to a normal user (like modification of the effective PATH).

Once this logic has been released at least once on GitHub, we can improve the test to have a real world upgrade scenario: for instance, first install the latest release, then try to upgrade to the dev version.

adferrand commented:
A nightly tests suite passing with these new tests, for the latest commit of this PR, is available here: https://dev.azure.com/adferrand/certbot/_build/results?buildId=433

bmw commented:
I'm having some trouble testing this. When I build an installer on your branch, install it, open Task Scheduler as an administrator and run the task, it errors out with "The operator or administrator has refused the request". Nothing is included about the task in the history tab.

I don't get this behavior when using our latest released installer. The tasks run and show up in the history. It does say that the last run result is "0x1" which may be a different problem.

adferrand commented:
The fact that I ran the installer on two machines and that the tests are passing in the CI may indicate that it it specific to your environment. Does any of the suggested root cause and solutions described here applies for you : https://stackoverflow.com/questions/36281197/the-operator-or-administrator-has-refused-the-request-task-scheduler

bmw commented:
I was able to get things mostly working. I don't think I really changed anything, but it's possible I was misreading things before. I still sometimes see "The operator or administrator has refused the request" in the "Last run result" column, but if I refresh it goes away.

I am still seeing the change where there is no history for the scheduled task shown in the task scheduler for the new task though. For tasks created by an old version of the installer, I see entries for launching the task, running "certbot renew", and finishing the task under the "history" tab. For the scheduled task in this PR, I see nothing. Is this an expected change? Do you see the history on your machines?

adferrand commented:
Ok. Now you will see again the task history ...

bmw commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    function Write-Error($message) {
+        Write-EventLog -Source "certbot/auto-update.ps1" -LogName "CertbotAutoUpdate" -EventID 1 -EntryType Error -Message $message
```
nit: Instead of defining `"certbot/auto-update.ps1"`, `"CertbotAutoUpdate"`, and maybe the event ID(?) in multiple places, what do you think about defining them earlier in the script as variables and reusing them here.

Quickly testing it, powershell seems to allow scoping like
```
$foo = "foo"
function Echo-Foo() {
    echo $foo
}
```

adferrand replied:
Done.

bmw commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+        Write-Host $message
+    }
+
+    function Write-Error($message) {
```
nit: Can we change the name of this function slightly to reflect that this also raises an exception? Maybe something like `Error`, `Fatal`, or `Exit`?

adferrand replied:
Done.

bmw commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+$_
+Assuming Certbot is not up-to-date.
+"@ "Warning"
+        $currentVersion = "0.0.0"
```
Should this also be a `System.Version` object?

adferrand replied:
Indeed!

adferrand commented:
I added in this PR the logic to check powershell version during the installation process. Do you want me to add also the powershell version check for the azure pipeline in this PR, or in another one @bmw?

bmw commented:
> Do you want me to add also the powershell version check for the azure pipeline in this PR, or in another one @bmw?

I don't have a strong preference, but if it's not significantly easier to do it here, I think I'd prefer it was done in another PR.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+
+    # Get latest remote certbot version
+    try {
+        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
```
Initially I added this line because of Windows 2012 R2: GitHub accepts only TLS 1.1+ and only TLS 1.0 was enabled for this system. At the basic level, then this line is not needed anymore because we do not support this system.

Now I think there is still something to do. At least we need to have the line `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12)`, because it will enable the use of TLS 1.2 at least on all systems, or a better TLS version if it is available.

Now if GitHub decides on day to accept only TLS 1.3, this code will not break either if:
* the systems have all TLS 1.3 available and use it by default
* or we update the current script to enable TLS 1.3 or better, with `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls13)`

However we cannot do anything for systems that do not have TLS 1.3 available.

I replace the line by `[Net.ServicePointManager]::SecurityProtocol = ([Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12)`.

bmw replied:
I looked into this a bit more and think we may still have a problem. According to https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.securityprotocol?view=netframework-4.8, since .NET 4.7, the default value of `[Net.ServicePointManager]::SecurityProtocol` is [SecurityProtocolType.SystemDefault](https://docs.microsoft.com/en-us/dotnet/api/system.net.securityprotocoltype?view=netframework-4.8#System_Net_SecurityProtocolType_SystemDefault) whose numerical value is 0. See:
```
PS /> [Net.ServicePointManager]::SecurityProtocol
SystemDefault
PS /> [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12
Tls12
```
This to me makes it look like it is not possible to add TLS 1.2 to the list of protocols the OS supports by default and trying to do so makes Powershell only support that protocol.

Unless I'm misunderstanding how this works or there is a better way to add TLS 1.2 to the list of protocols, I'm personally towards leaving this with its default value. What do you think?

adferrand replied:
Too bad. It was not working like this on my machine, and the value was an explicit composition of tls values. Well, better to remove that then, it works anyway out of the box for now with the supported Windows versions.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+    # Get latest remote certbot version
+    try {
+        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+        $result = Invoke-RestMethod -Uri $certbotUpgradeApiURL
```
Yes we should. I set 60s.

adferrand commented on windows-installer/template.nsi:
```diff
@@ -19,6 +19,17 @@ InstallDir "$PROGRAMFILES\Certbot"
 
+Function .onInit
+  ; Check that Powershell is at least 5.0.
+  nsExec::ExecToStack `powershell -ExecutionPolicy RemoteSigned -Command "Write-Host -NoNewline (Get-Host | Select-Object Version).Version.CompareTo([System.Version]'5.0')"`
```
As I said in the previous PR, the two approaches a very similar, and it is quite difficult to choose.

However here using `$PSVersionTable.PSVersion` would complexify the code, because the interpolation of `$VARIABLE` between NSIS and the underlying powershell code would need to be taken care of, and the comparison with the straight forward invocation of `[System.Version]'5.0'` is not possible anymore.

So I prefer to stay with `(Get-Host | Select-Object Version)` here.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,139 @@
+            }
+            # Install new version of Certbot
+            Write-Message "Running the installer for new version (install dir: $installDir) ..."
+            Start-Process -FilePath $installerPath -ArgumentList "/S /D=$installDir" -Wait
```
Invoking `/S` make the installer silent, and is specifically designed for non interactive calls. `-NoNewWindow` is not necessay here because the installer will not output anything, no logs or no UI.

And you have the answer for the second question: there is no logs to get here. The only outcome will be the exit code, that will trigger or not a global exception in the `auto-update.ps1` script (which is logged).

adferrand replied:
Noted!

adferrand commented on windows-installer/template-nsi.patch:
```diff
@@ -0,0 +1,183 @@
+--- pyapp.nsi	2020-03-31 01:46:13.118535700 +0200
```
How is a good old `diff`. What its usage becomes less clear now. Initally I wanted to keep track of all modifications done to the original file, but I realize that you can get them if you keep track of file source somewhere, and it is in the inline comments at the beginning of `template.nsi`.

So I suppose it is not useful anymore. I remove it.

adferrand replied:
I was not clear enough in my previous message: I wanted to keep track of modifications to dynamically change the original file during the installer compilation, but dropped the idea because it was too complex. So now it is useless.

adferrand commented on certbot-ci/windows_installer_integration_tests/test_main.py:
```diff
@@ -3,57 +3,211 @@
-        #                      '.UninstallString', capture_stdout=True)
-        # subprocess.check_call([uninstall_path, '/S'])
+        _ps('Remove-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotUpgradeApiURL')
+        _ps('Remove-ItemProperty -Path HKLM:\\Software\\Certbot -Name CertbotSigningPubKey')
```
Come on, having useless entries remain when the uninstaller kicks in, then having to use CCleaner to clean you registry after 6 months of that process, is a national sport for Windows users!

More seriously, deleting the entire `HKLM:\\Software\Certbot` is not that obvious. For instance a lot of programs use the registry to store user data, and I expect Certbot to do it soon also. In particular, we will offer the possibility to choose what is the default config folder to use. This value should not be deleted without explicit user approval, and this will be necessary during the auto upgrade process, since we uninstall Certbot first, so we will need to have a way to remember what was the user config folder in previous installation.

For these reasons, I think it is better to delete explicitly only the keys that relates to the installation of Certbot itself, and not the user, and I think it is good to start using this approach now.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+        try {
+            # Download the installer
+            Write-Message "Downloading the installer ..."
+            $webClient = New-Object System.Net.WebClient
```
Indeed, we can reuse Invoke-RestMethod here. I am not convinced by the wrapper for two calls, since it requires boilerplate code to handle the very different situations here (REST API call and binary download). I added a timeout however.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+        # Search for the Windows installer asset
+        $installerUrl = $null
+        foreach ($asset in $result.assets) {
+            if ($asset.name -match '^certbot-.*installer-win32\.exe$') {
```
I made the installer name a constant in `construct.py`, and added inline comments.

adferrand commented on windows-installer/auto-update.ps1:
```diff
@@ -0,0 +1,134 @@
+
+        Write-Message "Starting Certbot auto-upgrade from $currentVersion to $latestVersion ..."
+
+        $installerPath = "$env:TMP/certbot-installer-win32.exe"
```
Here is a better implementation. It uses a high level .NET function to get the user TMP path, and store the installer in a temporary directory with random name to avoid collisions. For the rest, the user TMP path is accessible only to the user, so in terms of DACL it is secure.

adferrand commented on certbot-ci/windows_installer_integration_tests/test_main.py:
```diff
@@ -3,57 +3,211 @@
+        class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
+            pass
+
+        class GitHubMock(BaseHTTPRequestHandler):
```
Not possible for GitHubMock. The `do_GET` method is constructed with a closure on the installer variable exposed by the current instance of the fixture.

I move ThreadedTCPServer outside.

adferrand commented on certbot-ci/windows_installer_integration_tests/test_main.py:
```diff
@@ -3,57 +3,211 @@
+                            'tag_name': 'v99.9.9',
+                            'assets': [{
+                                'name': os.path.basename(installer),
+                                'browser_download_url': 'http://localhost:8009/{0}'.format(os.path.basename(installer))
```
I suppose here that you are refering to the value of the port. Done.

adferrand commented on certbot-ci/windows_installer_integration_tests/test_main.py:
```diff
@@ -3,57 +3,211 @@
+import pytest
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    import pkg_resources
```
I fixed that by upgrading setuptools during the Azure pipeline execution.

adferrand commented on .azure-pipelines/templates/installer-tests.yml:
```diff
@@ -49,6 +49,7 @@ jobs:
         displayName: Retrieve Windows installer
       - script: |
           py -3 -m venv venv
+          venv\Scripts\python -m pip install --upgrade pip setuptools
```
Yes, I remembered. I use pipstrap now.

adferrand commented on certbot-ci/windows_installer_integration_tests/test_main.py:
```diff
@@ -59,19 +58,20 @@ def installer(request, signing_cert):
 def github_mock(installer):
     """
-    This fixture starts a GitHub release API mock on localhost:9001. This mock returns a
+    This fixture starts a GitHub release API mock on localhost:8009. This mock returns a
```
Done.

bmw commented:
I'm not sure what's going on with the failures to install `pywin32==227`. If you want, I think it's OK to just pin `pip` and `setuptools` inline to a newer version which apparently do not have this problem.

Other than that being fixed, this PR LGTM! The two things I can think of that we need before merging it are:

- [ ] Updating the public key in this PR to match the key in our certificate
- [ ] Instructions/tools for making this signature. I think we should try to make this work on macOS/Linux as having to spin up a Windows machine manually during the release is a little painful. It looks like https://github.com/mtrojnar/osslsigncode might work for this. We will probably want to store the private key on a smart card so we should try to figure out how to make that work. Docs like https://github.com/larskanis/osslsigncode/blob/39a4a5032f07a99f497322ce002e417e5c03c983/README#L107-L113 seem to suggest this is possible with `osslsigncode`.

bmw commented:
Using the instructions in my last post about `osslsigncode`, I was able to create a valid authenticode signature using a dummy cert and key I put on my smart card on macOS.

I'm still open to other options if we find something better, but it seems to me like that'll work!

bmw commented:
We may want to think about changing the architectures we support for Windows as discussed briefly at https://github.com/certbot/certbot/issues/8315 before we release this feature. I think doing it afterwards is more complicated because we may have multiple installers to choose from or we may drop support for 32-bit systems.

bmw commented:
I think we should start thinking about https://github.com/certbot/certbot/issues/8315. One option is convincing ourselves that we're able to easily make changes here afterwards or we could go ahead and tackle that issue now.

Also, we're planning on going ahead with the plan at https://github.com/certbot/certbot/issues/8047#issuecomment-740881609 so this PR will need to be updated slightly to trust 3 different keys (which have not been generated yet).

zoracon commented:
Closing due to our decision to discontinue our Beta distribution of the Windows Installer:
https://github.com/certbot/certbot/wiki/Architectural-Decision-Records#-certbot-discontinuing-windows-support

If you need alternatives. They are listed in this community forum post:
https://community.letsencrypt.org/t/certbot-discontinuing-windows-beta-support-in-2024/208101
